File: LeetCode/322-coin-change/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        memo = {}

        def dp(amount):
            if amount == 0:
                return 0

            retList = []
            for coin in coins:
                remain_amount = amount - coin
                if remain_amount >= 0:
                    if remain_amount not in memo:
                        memo[remain_amount] = dp(remain_amount)
                    if memo[remain_amount] != -1:
                        retList.append(memo[remain_amount])
            return min(retList) + 1 if retList else -1

        return dp(amount)

File: LeetCode/322-coin-change/coin-change.java
class Solution {
    private Map<Integer, Integer> memo = new HashMap<>();

    public int coinChange(int[] coins, int amount) {
        return dp(coins, amount);
    }

    private int dp(int[] coins, int amount) {
        if (amount == 0) return 0;
        if (memo.containsKey(amount)) return memo.get(amount);

        int minCount = Integer.MAX_VALUE;
        for (int coin : coins) {
            int remainAmount = amount - coin;
            if (remainAmount >= 0) {
                int result = dp(coins, remainAmount);
                if (result != -1) {
                    minCount = Math.min(minCount, result + 1);
                }
            }
        }

        int result = (minCount == Integer.MAX_VALUE) ? -1 : minCount;
        memo.put(amount, result);
        return result;
    }
}